### PR TITLE
`Paywalls`: send events to `Purchases`

### PR DESCRIPTION
**Pull request by NachoSoto**
Last part, follow up to #3160.

This now makes `PaywallView` track events through the `RevenueCat` SDK.

### CODE REVIEW

File: RevenueCatUI/Purchasing/PurchaseHandler.swift
Comment (NachoSoto) [inline RevenueCatUI/Purchasing/PurchaseHandler.swift:156]: This is covered by tests. I tried different approaches of where to create the session ID to avoid this, but this was the simplest.

File: RevenueCatUI/Views/LoadingPaywallView.swift
Comment (NachoSoto) [inline RevenueCatUI/Views/LoadingPaywallView.swift:76]: I love that the compiler made me implement this. And realized this was the best solution.

File: Tests/RevenueCatUITests/PaywallEventsTests.swift
Comment (NachoSoto) [inline Tests/RevenueCatUITests/PaywallEventsTests.swift:24]: These tests cover everything 🎉

File: RevenueCatUI/Data/TestData.swift
Comment (aboedo) [inline RevenueCatUI/Data/TestData.swift:190]: this quickly became a BIG file. Maybe we should move it to a folder and split into smaller things?

Comment (aboedo) [reply RevenueCatUI/Data/TestData.swift:190]: it'll only grow over time

File: RevenueCatUI/PaywallView.swift
Comment (aboedo) [inline RevenueCatUI/PaywallView.swift:240]: does onDisappear get called when the app gets backgrounded? And when the iOS payment window pops up?

Comment (aboedo) [reply RevenueCatUI/PaywallView.swift:240]: wondering whether we need to [do something like this](https://lyons.app/2020/08/31/how-to-only-run-ondisappear-when-a-view-is-unloaded-in-swiftui/#:~:text=How%20to%20only%20run%20onDisappear%20when%20a%20view,code%20when%20the%20view%20is%20dismissed%20or%20unloaded.)

File: RevenueCatUI/PaywallView.swift
Comment (aboedo) [inline RevenueCatUI/PaywallView.swift:271]: Note: we're baking in an assumption here: 
We're saying that if SwiftUI keeps the PaywallView in memory but calls onAppear more than once, we consider it to be the same session, right? 
If that's the case, maybe we should explicitly call it out here?

File: RevenueCatUI/Purchasing/PurchaseHandler.swift
Comment (aboedo) [inline RevenueCatUI/Purchasing/PurchaseHandler.swift:109]: I'm super late to this, but given that we already use "view" and "paywallView" a bunch because SwiftUI, I'd call this a Paywall Impression event

File: RevenueCatUI/Purchasing/PurchaseHandler.swift
Comment (aboedo) [reply RevenueCatUI/Purchasing/PurchaseHandler.swift:156]: Could you elaborate on what problem this approach solves? I'm trying to understand how we ended up with trackEvent taking a function that turns data into an event, and I'm sure there's a reasonable explanation but Fried Brain Fridays

File: RevenueCatUI/Data/TestData.swift
Comment (NachoSoto) [reply RevenueCatUI/Data/TestData.swift:190]: Yup 👍🏻  I'll do it next week.

File: RevenueCatUI/Purchasing/PurchaseHandler.swift
Comment (NachoSoto) [reply RevenueCatUI/Purchasing/PurchaseHandler.swift:109]: Good idea 👍🏻

File: RevenueCatUI/Purchasing/PurchaseHandler.swift
Comment (NachoSoto) [reply RevenueCatUI/Purchasing/PurchaseHandler.swift:156]: This method relies on `self.eventData`, because that needs to be saved throughout the lifetime of a `PaywallView` (since it has the session ID, dark mode, locale, etc).
The closure parameter is just a way to avoid duplicated code. It takes `self.eventData` (which is a `PaywallEvent.Data`) and wraps it into a `PaywallEvent`.
So it gets used like this:
```
self.trackEvent(PaywallEvent.view)
self.trackEvent(PaywallEvent.cancel)
self.trackEvent(PaywallEvent.close)
```

File: RevenueCatUI/PaywallView.swift
Comment (NachoSoto) [reply RevenueCatUI/PaywallView.swift:240]: Good point, let me test this + that approach.

File: RevenueCatUI/Purchasing/PurchaseHandler.swift
Comment (aboedo) [reply RevenueCatUI/Purchasing/PurchaseHandler.swift:156]: it took me a few mins to understand the implementation tbh. I got confused with the closure, the fact that there are a few things called trackEvent and the mapping on the init. 

Would this work? 

- always pass in the full paywall data for events, including cancel and close. The event reporting layer has that data after all, and that protects against unexpected cases where cancel comes before view or something (we're delegating into background priority tasks, and although I imagine things happen in order, in case of a freak accident maybe it'd be safer to always have the data). 
- remove the closures, making the implementation of this method something like
```swift
    func track(event: PaywallEvent) {
        Task.detached(priority: .background) { [block = self.trackEventBlock] in
            await purchases.track(paywallEvent: event)
        }
    }
```

and called like 
```swift
func trackPaywallView(_ eventData: PaywallEvent.Data) {
    self.track(event: PaywallEvent.view(eventData))
}
...
self.purchaseHandler.trackPaywallView(self.eventData)
```

Comment (aboedo) [reply RevenueCatUI/Purchasing/PurchaseHandler.swift:156]: we'd need to also pass in the paywall data when calling purchase and restore, but I think that'd be fine? the caller knows the paywall information anyway

Comment (NachoSoto) [reply RevenueCatUI/Purchasing/PurchaseHandler.swift:156]: Refactored

File: RevenueCatUI/PaywallView.swift
Comment (NachoSoto) [reply RevenueCatUI/PaywallView.swift:271]: Yup, I'll document it (I've also confirmed with @joshdholtz that repeated view events will be handled in the backend).

File: RevenueCatUI/Purchasing/PurchaseHandler.swift
Comment (NachoSoto) [reply RevenueCatUI/Purchasing/PurchaseHandler.swift:109]: I'll make a PR renaming this in the code. I'll keep the backend event identifier as "view" unless @joshdholtz is also okay changing that.

File: RevenueCatUI/PaywallView.swift
Comment (NachoSoto) [reply RevenueCatUI/PaywallView.swift]: The answer to both is no :) so the simple approach here works perfectly without extra hacks :)

File: RevenueCatUI/PaywallView.swift
Comment (aboedo) [inline RevenueCatUI/PaywallView.swift:227]: ❤️

File: RevenueCatUI/Purchasing/PurchaseHandler.swift
Comment (aboedo) [inline RevenueCatUI/Purchasing/PurchaseHandler.swift:23]: These are now only needed for testing purposes, right? 
Is it worth keeping them around for that? Having them also forces us to have fatalError workarounds in the loading view IIRC?

Would it suffice to have a purchases mock (spy) and pass that in, so that we can directly call the methods from purchases? Seems like we only use them to check that we're calling the right methods with the right params

Comment (NachoSoto) [reply RevenueCatUI/Purchasing/PurchaseHandler.swift:23]: These serve several purposes:
- Abstract implementation from `Purchases`
- Previews: we can have custom handlers that don't do anything or that complete a successful purchase
- Snapshots: same thing
- Unit tests (`PaywallFooterTests` for example)
- `LoadingPaywallView`: this one doesn't use real blocks for obvious reasons

> Having them also forces us to have fatalError workarounds in the loading view IIRC?

We have a `fatalError` in case you try to purchase from the loading view (which is `.disabled`, so it's impossible). Other than that, the `trackEvent` one is a no-op now.

Hopefully that answers your question why we can't just use a `Purchases` instance. We could create a `PurchasesType` implementation, but that would require implementing dozens of methods instead of just 3.

Comment (aboedo) [reply RevenueCatUI/Purchasing/PurchaseHandler.swift:23]: Yeah, that's what I was getting at, essentially for DI. 

I don't feel strong enough about this to continue to block this PR, but I do think there are alternative approaches worth studying, that wouldn't necessitate building a mock of all methods in `PurchasesType` from scratch. 

I worry about us adding a different approach for testing (instead of DI of a class, DI at the closure level). It makes things more complex for a brand new maintainer. And I'd argue that the closures are harder to read and wrap your head around than just passing in a purchases object, although that bit is subjective. 

Alternate approaches that I think would also do the trick: 
- find a way to share the purchases mock instance we use for other tests. We already share some tests code, this wouldn't fall too far from that
- create a new, much smaller protocol, that only has the methods that we'd need (purchase, restore, track events). Refactor PurchasesType so that it also conforms to this protocol. Then we pass in an instance of this smaller protocol, which is easy to mock. Then PurchasesHandler just takes a `purchases: NewSmallType = Purchases.shared` as param, and the rest of the class is dead simple.

**Review by aboedo — APPROVED** [review]

File: RevenueCatUI/Purchasing/PurchaseHandler.swift
Comment (NachoSoto) [reply RevenueCatUI/Purchasing/PurchaseHandler.swift:23]: Thanks for clarifying what you meant! It does indeed lead to a simpler implementation :)

Done in #3196

File: RevenueCatUI/Data/TestData.swift
Comment (NachoSoto) [reply RevenueCatUI/Data/TestData.swift:190]: Also done in #3196

**Pull request merged**
